frameless-runtime/src/utxo.rs

```diff
@@ -0,0 +1,253 @@
+#[derive(PartialEq, Eq, PartialOrd, Ord, Default, Clone, Encode, Decode, Hash, Debug, TypeInfo)]
+pub struct Input {
+    /// A previously created output that will be consumed by the transaction containing this input.
+    pub output: OutputRef,
+    /// A witness proving that the output can be consumed by this input. In many cases including that of a basic cryptocurrency, this will be a digital signature.
+    pub witness: Sig,
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, parity_util_mem::MallocSizeOf))]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Default, Clone, Encode, Decode, Hash, Debug, TypeInfo)]
+pub struct Output {
+    /// The address that owns this output. Based on either a public key or a Tuxedo Piece
+    pub redeemer: Redeemer,
+    /// The data associated with this output. In the simplest case, this will be a token balance, but could be arbitrarily rich state.
+    pub data: Value,
+    /// An Id for this type Such that we know how to encode or decode the 'data' field
+    pub data_id: TypeId,
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, parity_util_mem::MallocSizeOf))]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Default, Clone, Encode, Decode, Hash, Debug, TypeInfo)]
+pub struct Transaction {
+    /// The inputs refer to currently existing unspent outputs that will be consumed by this transaction
+    pub inputs: Vec<Input>,
+    /// Similar to inputs, Peeks refer to currently existing utxos, but they will be read only, and not consumed
+    pub peeks: Option<Vec<Input>>,
+    /// The new outputs to be created by this transaction.
+    pub outputs: Vec<Output>,
+}
+
+pub type Utxo = Output;
+pub type UtxoRef = OutputRef;
+
+pub trait Redeem {
+    fn redeem(self, tx: &[u8], witness: &[u8]) -> bool;
+}
+
+impl Redeem for Redeemer {
+    fn redeem(self, tx: &[u8], witness: &[u8]) -> bool {
+        let signature = match Signature::try_from(&witness[..]) {
+            Ok(sig) => sig,
+            Err(_) => return false,
+        };
+        sp_io::crypto::sr25519_verify(&signature, &tx, &Public::from_h256(self))
+    }
+}
+
+// TODO: Implement this for Each Tuxedo Piece
+pub trait UtxoSet {
+
+    /// TODO: Change these bool return types to Result types for more error propagation clarity
+
+    /// Check whether a given utxo exists in the current set
+    fn contains(utxo_ref: UtxoRef) -> bool;
+
+    /// Insert the given utxo into the state storing it with the given ref
+    /// The ref is probably the hash of a tx that created it and its index in that tx, but this decision is opaque to this trait
+    /// Return whether the operation is successful (It can fail if the ref is already present)
+    fn insert(utxo_ref: UtxoRef, utxo: Utxo) -> bool;
+
+    ///
+    /// nullify the utxo by either:
+    /// - Consuming it entirely
+    /// - Putting it on "Timeout"
+    /// - Not consuming it but marking it as spent
+    ///
+    fn nullify(utxo_ref: UtxoRef) -> Option<Utxo>;
```

Comment:
I like the flexibility you're planning for here. We might need to add another parameter to the `nullify` function (possibly with a generic type). For example, in the ZK utxo set, we would have to provide a nullifier (like publishing the serial number) right? I guess in the simple case of a totally transparent utxo set that type could just be `()`. Just thinking out loud here.

Reply:
Would like to explore this thought a bit more again together then ask Jeff about it